pylint/checkers/base/function_checker.py

```diff
@@ -0,0 +1,163 @@
+
+            from contextlib import contextmanager
+
+            @contextmanager
+            def cm():
+                print("cm enter")
+                a = yield
+                print('cm exit')
+
+            def genfunc():
+                with cm():
+                    print("stepping")
+                    yield
+
+            def main():
+                gen = genfunc()
+                ref = weakref.ref(gen, print)
+                next(gen)
+
+        The function ``cm`` needs to include error handling for ``GeneratorExit``
+        to properly finish the function execution.
+
+        :param node: FunctionDef node to check
+        :type node: nodes.FunctionDef
+        """
+        contextmanager_str = "contextmanager"
+        if node.decorators is None:
+            return
+
+        decorator_list = node.decorators.nodes
+        if not any(
+            contextmanager_str == self._get_name(deco) for deco in decorator_list
+        ):
+            # No contextmanager decorator
+            return
+
+        # A ``yield`` statement can either be Expr or Assign, but the value is always of Yield
+        # Additionally a good function could be input where the yield is already wrapped by a Try
+        # but not have the required Handler or Finally
+        body_nodes = self._find_yield_nodes(node)
+        yield_nodes: list[nodes.Yield] = []
+        for n in body_nodes:
+            if not isinstance(n, nodes.Try):
+                yield_nodes.append(n)
+                continue
+            # This Try Node has a Yield inside of it, making this a generator function wrapped
+            # by contextlib.contextmanager. See if handlers or finally is present
+            if (
+                any(
+                    self._get_name(handler.type) == "GeneratorExit"
+                    for handler in n.handlers
+                )
+                or n.finalbody
+            ):
+                continue
+
+            # this Try block doesn't properly handle GeneratorExit, add warning
+            self.add_message(
+                "contextmanager-generator-missing-cleanup", node=node, args=(node.name,)
+            )
+            return
+
+        if yield_nodes:
+            self.add_message(
+                "contextmanager-generator-missing-cleanup", node=node, args=(node.name,)
+            )
+
+    @staticmethod
+    def _get_name(node: nodes.NodeNG) -> str:
+        """Helper function to get the name from a node.
+
+        Helpful when possible to have a Name or Attribute node.
+
+        :param node: Node to get name of
+        :type node: nodes.NodeNG
+        :return: string name
+        :rtype: str
+        """
+        if hasattr(node, "name"):
+            return node.name  # type: ignore[no-untype-def]
+        if hasattr(node, "attrname"):
+            return node.attrname  # type: ignore[no-untype-def]
+        return node.as_string()  # type: ignore[no-untype-def]
```

Comment:
We have this problem approximately everywhere and generally we check the node type with ``isinstance``.